src/azul/service/source_service.py

```diff
@@ -68,10 +68,21 @@ def list_source_ids(self,
         assert not any(joiner in c for c in cache_key), cache_key
         cache_key = joiner.join(cache_key)
         try:
-            source_ids = set(self._get(cache_key))
+            source_ids = self._get(cache_key)
         except CacheMiss:
             source_ids = plugin.list_source_ids(authentication)
             self._put(cache_key, list(source_ids))
+        else:
+            try:
+                source_ids = set(source_ids)
+            except TypeError as e:
+                # FIXME: Remove safety net from source cache format transition
+                #        https://github.com/DataBiosphere/azul/issues/5204
+                if e.args == ("unhashable type: 'dict'",):
+                    source_ids = plugin.list_source_ids(authentication)
+                    self._put(cache_key, list(source_ids))
```

Comment:
Moving it to the bottom of the method causes a round-trip to TDR to be performed even on a valid cache hit. I came up a different way to avoid the duplication.

Reply:
What I meant is this:

```diff
Index: src/azul/service/source_service.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/service/source_service.py b/src/azul/service/source_service.py
--- a/src/azul/service/source_service.py	(revision c823f8c4c726312270d01d1684742f77e142d727)
+++ b/src/azul/service/source_service.py	(date 1684285225754)
@@ -70,19 +70,17 @@
         try:
             source_ids = self._get(cache_key)
         except CacheMiss:
-            source_ids = plugin.list_source_ids(authentication)
-            self._put(cache_key, list(source_ids))
+            pass
         else:
+            # FIXME: Remove safety net from source cache format transition
+            #        https://github.com/DataBiosphere/azul/issues/5204
             try:
-                source_ids = set(source_ids)
+                return set(source_ids)
             except TypeError as e:
-                # FIXME: Remove safety net from source cache format transition
-                #        https://github.com/DataBiosphere/azul/issues/5204
-                if e.args == ("unhashable type: 'dict'",):
-                    source_ids = plugin.list_source_ids(authentication)
-                    self._put(cache_key, list(source_ids))
-                else:
-                    raise
+                if e.args != ("unhashable type: 'dict'",):
+                    raise
+        source_ids = plugin.list_source_ids(authentication)
+        self._put(cache_key, list(source_ids))
         return source_ids
 
     def list_sources(self,

```

That diff is meant as an alternative to your fixup commit.